extensions/amp-bind/0.1/bind-expression.js

```diff
@@ -14,33 +14,253 @@
+          return null;
+        }
+        // Ignore Closure's type constraint for `hasOwnProperty`.
+        if (Object.prototype.hasOwnProperty.call(
+              /** @type {Object} */ (target), member)) {
+          return target[member];
+        }
+        return null;
+
+      case AstNodeType.MEMBER:
+        return value || this.eval_(args[0], scope);
+
+      case AstNodeType.VARIABLE:
+        const variable = value;
+        if (Object.prototype.hasOwnProperty.call(scope, variable)) {
+          return scope[variable];
+        }
+        return null;
+
+      case AstNodeType.ARGS:
+      case AstNodeType.ARRAY_LITERAL:
+        return (args.length > 0) ? this.eval_(args[0], scope) : [];
+
+      case AstNodeType.ARRAY:
+        return args.map(element => this.eval_(element, scope));
+
+      case AstNodeType.OBJECT_LITERAL:
+        return (args.length > 0)
+            ? this.eval_(args[0], scope)
+            : Object.create(null);
+
+      case AstNodeType.OBJECT:
+        const object = Object.create(null);
+        args.forEach(keyValue => {
+          const {k, v} = this.eval_(keyValue, scope);
+          object[k] = v;
+        });
+        return object;
+
+      case AstNodeType.KEY_VALUE:
+        return {
+          k: this.eval_(args[0], scope),
+          v: this.eval_(args[1], scope),
+        };
+
+      case AstNodeType.NOT:
+        return !this.eval_(args[0], scope);
+
+      case AstNodeType.UNARY_MINUS:
+        return -Number(this.eval_(args[0], scope));
+
+      case AstNodeType.UNARY_PLUS:
+        return +Number(this.eval_(args[0], scope));
+
+      case AstNodeType.PLUS:
+        return this.eval_(args[0], scope) + this.eval_(args[1], scope);
+
+      case AstNodeType.MINUS:
+        return Number(this.eval_(args[0], scope)) -
+            Number(this.eval_(args[1], scope));
+
+      case AstNodeType.MULTIPLY:
+        return Number(this.eval_(args[0], scope)) *
+            Number(this.eval_(args[1], scope));
+
+      case AstNodeType.DIVIDE:
+        return Number(this.eval_(args[0], scope)) /
+            Number(this.eval_(args[1], scope));
+
+      case AstNodeType.MODULO:
+        return Number(this.eval_(args[0], scope)) %
+            Number(this.eval_(args[1], scope));
+
+      case AstNodeType.LOGICAL_AND:
+        return this.eval_(args[0], scope) && this.eval_(args[1], scope);
+
+      case AstNodeType.LOGICAL_OR:
+        return this.eval_(args[0], scope) || this.eval_(args[1], scope);
+
+      case AstNodeType.LESS_OR_EQUAL:
+        return this.eval_(args[0], scope) <= this.eval_(args[1], scope);
+
+      case AstNodeType.LESS:
+        return this.eval_(args[0], scope) < this.eval_(args[1], scope);
+
+      case AstNodeType.GREATER_OR_EQUAL:
+        return this.eval_(args[0], scope) >= this.eval_(args[1], scope);
+
+      case AstNodeType.GREATER:
+        return this.eval_(args[0], scope) > this.eval_(args[1], scope);
+
+      case AstNodeType.NOT_EQUAL:
```

Comment:
Currently we support only abstract equality for simplicity. Bind expressions aim to be a user-friendly and JS-*like*, not necessarily 100% faithful.